File: 04-Data-Structures/code/typescript/src/stack.ts
import { DSError } from "./errors.js";
import { DynamicArray } from "./dynamicArray.js";

/** LIFO stack over a DynamicArray. */
export class Stack<T> {
  private arr = new DynamicArray<T>();

  push(x: T): void {
    this.arr.push(x);
  }

  pop(): T {
    if (this.arr.size() === 0) throw new DSError("empty", "pop from empty stack");
    return this.arr.pop();
  }

  peek(): T {
    if (this.arr.size() === 0) throw new DSError("empty", "peek from empty stack");
    return this.arr.get(this.arr.size() - 1);
  }

  size(): number {
    return this.arr.size();
  }

  isEmpty(): boolean {
    return this.arr.size() === 0;
  }

  checkInvariants(): void {
    this.arr.checkInvariants();
  }
}

File: 04-Data-Structures/code/python/seb_ds/stack.py
"""LIFO stack over a DynamicArray. Mirrors `typescript/src/stack.ts`."""

from __future__ import annotations

from typing import Generic, TypeVar

from .dynamic_array import DynamicArray
from .errors import DSError

T = TypeVar("T")


class Stack(Generic[T]):
    def __init__(self) -> None:
        self._arr: DynamicArray[T] = DynamicArray()

    def push(self, x: T) -> None:
        self._arr.push(x)

    def pop(self) -> T:
        if self._arr.size() == 0:
            raise DSError("empty", "pop from empty stack")
        return self._arr.pop()

    def peek(self) -> T:
        if self._arr.size() == 0:
            raise DSError("empty", "peek from empty stack")
        return self._arr.get(self._arr.size() - 1)

    def size(self) -> int:
        return self._arr.size()

    def is_empty(self) -> bool:
        return self._arr.size() == 0

    def check_invariants(self) -> None:
        self._arr.check_invariants()
